Add bswap built-in function

gedimitr opened the pull request:
The function takes an unsigned integer and returns an unsigned
integer of the same width, but with the byte order reversed.

Resolves #1875 

One thing I'd like to draw your attention is the integer type promotion / demotion in order to fix a strange behaviour when 16 bit arguments were received. Since I am not really competent in LLVM, it is possible that something is escaping me....

Also, I have updated TCP related programs (e.g. `tcpconnect.bt`) to use the new bswap function, instead of the manual byte reordering.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

fbs commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -1088,6 +1088,31 @@ void CodegenLLVM::visit(Call &call)
+      // correct bswap intrinsic would be selected, (bswap.i16), yet LLVM would
+      // give to it a zero extended i64 argument! The trick below seems to fix
+      // this behaviour and the argument to bswap is always of the correct type.
+      Value *promoted_value = b_.CreateIntCast(expr_, b_.getInt64Ty(), false);
```
does this happen in the original IR or the optimized one? We have some weird casts in there for 'reasons' which can make the optimized ir confusing

fbs commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -1088,6 +1088,31 @@ void CodegenLLVM::visit(Call &call)
+    auto scoped_del_arg = accept(arg);
+
+    assert(arg->type.IsIntegerTy());
+    if (arg->type.GetSize() > 1)
```
ah yeah skipping the byte case makes sense.

re assert, we should try to avoid them as the end user experience of asserts is quite bad. But in this case have test coverage for it (`test("kprobe:f { bswap(\"hello\"); }", 10);`) so we will probably catch it before it ends up in the master branch

gedimitr commented:
I did a more throrough investigation and I added a changeset which fixes this strange behaviour:
* When an element was read from a compound data structure in `probereadDatastructElem`, the returned value would uncoditionally be cast to a 64 bit integer
* By removing this cast, the narrower types get correctly handled
* The promoting / demoting trick is no longer necessary

After running the runtime tests, some errors were reported for some test cases due to differences in the generated IRs. I had a look into the diffs and they appear to be benign.

I also added a fix in the semantic analysis of bswap, because assertion failures were occurring when the tests were running in debug builds.

fbs commented:
> I did a more throrough investigation and I added a changeset which fixes this strange behaviour:
> 
>     * When an element was read from a compound data structure in `probereadDatastructElem`, the returned value would uncoditionally be cast to a 64 bit integer
> 
>     * By removing this cast, the narrower types get correctly handled
> 
>     * The promoting / demoting trick is no longer necessary
> 
> 

Before we merge this we should run all the tools we have and make sure the output makes sense. We had some issues in the past and these weird casts are part of it, e.g. #1173

gedimitr commented:
Went through all the programs in the `tools` directory and run them: all of them executed successfully and provided sensible results.

Exceptions (that by all accounts are irrelevant to the changes introduced with this PR):
- I got some warnings (e.g. `biolatency.bt` complained that `__blk_account_io_start` / `__blk_account_io_done` is not traceable) 
- `mdflush` failed to start due to an error in a `struct` definition (ERROR: Struct/union of type 'struct bio' does not contain a field named 'bi_bdev')
- The `xfsdist.bt` program wasn't tested for results accuracy, since no XFS filesystem was available (the program was successfully started and the probes established)

viktormalik commented:
> Exceptions (that by all accounts are irrelevant to the changes introduced with this PR):
> 
> * I got some warnings (e.g. `biolatency.bt` complained that `__blk_account_io_start` / `__blk_account_io_done` is not traceable)
> * `mdflush` failed to start due to an error in a `struct` definition (ERROR: Struct/union of type 'struct bio' does not contain a field named 'bi_bdev')
> * The `xfsdist.bt` program wasn't tested for results accuracy, since no XFS filesystem was available (the program was successfully started and the probes established)

I can confirm that these should be unrelated to this PR.

viktormalik reviewed (commented):
LGTM, thanks!

I think that getting rid of that unconditional cast is reasonable. Looking at the changes in codegen tests, it seems that each operation does typecasts on its inputs when needed (which is the way it should be done anyway), so hopefully nothing will break.

Could you update the codegen tests so that CI passes? There's a script in `scripts/` for that.

viktormalik reviewed (approved):
Great, thanks! Could you squash the commits, please?

gedimitr commented:
@viktormalik Just pushed everything as a single comit. Thanks!
